### PR TITLE
consolidate RouteRequestCallback and Router.Callback into a single RouterCallback with improved error accessibility

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

Refs https://github.com/mapbox/mapbox-navigation-android/issues/4376. This contribution refactors our route requests callback, removes the duplication in the public API and exposes much more information about the reason for the failure, including the original request URL and HTTP codes and error bodies (when available).

In short, we're conslidating:
```kotlin
interface Router.Callback {
    fun onResponse(routes: List<DirectionsRoute>)
    fun onFailure(throwable: Throwable)
    fun onCanceled()
}
```
and
```kotlin
interface RoutesRequestCallback {
    fun onRoutesReady(routes: List<DirectionsRoute>)
    fun onRoutesRequestFailure(throwable: Throwable, routeOptions: RouteOptions)
    fun onRoutesRequestCanceled(routeOptions: RouteOptions)
}
```
that were used in different parts of the codebase (and ultimately did the same thing) into:
```kotlin
interface RouterCallback {
    fun onRoutesReady(routes: List<DirectionsRoute>)
    fun onFailure(reasons: List<RouterFailure>, routeOptions: RouteOptions)
    fun onCanceled(routeOptions: RouteOptions)
}
```

where `RouterFailure` provides a code and reason, also for nested failures.
```kotlin
data class RouterFailure(
    val url: URL,
    val message: String,
    val code: Int? = null,
    val throwable: Throwable? = null
)
```
For example, if both offboard and onboard routers fail, we now have a way to provide both reasons.

The next steps are to expose the [Directions API error codes](https://docs.mapbox.com/api/navigation/directions/#directions-api-errors) as string defs or enums to fully close https://github.com/mapbox/mapbox-navigation-android/issues/4376, and make the similar refresh interfaces internal and hidden since they are not used outside of the SDK scope.

### Changelog

```
<changelog>:warning: Removed `RouteRequestCallback` and `Router.Callback` in favor of a single `RouterCallback` with improved failure reason accessibility.</changelog>
```

### CODE REVIEW

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt
Comment (korshaknn) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:91]:
```suggestion
                                RouterFailure(url, response.errorBody()?.string()), code)
```
can be simplified?

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt
Comment (korshaknn) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:107]: move  "unknown" to constants

File: libnavigation-router/src/test/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouterTest.kt
Comment (korshaknn) [inline libnavigation-router/src/test/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouterTest.kt:227]: there is a global 
```
private val url = HttpUrl.Builder()
        .scheme("https")
        .host("api.mapbox.com")
        .addPathSegment("test")
        .build()
```
can the local be removed?

**Review by korshaknn — COMMENTED** [review]
great work! added some comments

File: examples/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt
Comment (LukasPaczos) [reply examples/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt:560]: Good catch, I was testing something else :upside_down_face:

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:91]: `ResponseBody#string` is nullable, that's why we need the extension

**Comment by LukasPaczos** [comment]
Ready for another round.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt:108]: Do we really want to log out now that the `reasons` are available? 🤔 

Same thing applies to `onCanceled`.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: In any case, we're already logging in `MapboxDirectionsSession`, do we want to log again here?

Same for `onCanceled` 👇

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/RerouteController.kt
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/RerouteController.kt:105]:
```suggestion
    data class Failed @JvmOverloads constructor(
```

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/hybrid/HybridRouterHandler.kt
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/hybrid/HybridRouterHandler.kt:58]: As mentioned above, we're also logging here. Do we need all this logging "duplication"?

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:89]: Didn't we agree that this is not an option? cc @danpat @mskurydin @vanmaxim

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:74]: For "security" reasons, do we want to filter out the `access_token` from the URL so it's not leaked out in the Logcat?

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/ElapsedTime.kt
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/ElapsedTime.kt:19]:
```suggestion
            throw IllegalStateException("Must call start() before calling end()")
```

**Review by Guardiola31337 — COMMENTED** [review]
Great work @LukasPaczos 

Left some comments to discuss before merging here.

**Comment by Guardiola31337** [comment]
@LukasPaczos 
> make the similar refresh interfaces internal and hidden since they are not used outside of the SDK scope

You mean refactor _Refresh_ APIs similarly to `Router` ones and provide same consistent messages and failures, right?

**Comment by LukasPaczos** [comment]
I meant moving the refresh interfaces from `base` to `core` and making them internal (as much as possible), or just moving them to `internal` package and outside of the public API. Those interfaces are only used internally. We can refactor them similarly though so that we can log out the refresh failure better.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:74]: :+1:

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt:108]: Probably it's not necessary anymore :+1:

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt:74]: If we won't log this out by default, I think adding a separate field makes sense :+1:

**Comment by Guardiola31337** [comment]
Yeah that makes total sense 👍 

> We can refactor them similarly though so that we can log out the refresh failure better.

Yeah, for consistency would like to do this too (not blocking this PR though) 👍

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt:108]: Alternatively, we can decrease the log level to keep this around by default but not pollute the top log levels as much.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt:74]: Let me follow up on online/offline changes in a separate PR to not overload this one. We can then add the flag to the failure reason as well.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: I think this one should stay. Developers can't hook into a failure of our default reroute controller, so we need to have a way to communicate the errors.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: Ah this is not true, developers can use `mapboxNavigation.getRerouteController()?.registerRerouteStateObserver` :+1:

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/hybrid/HybridRouterHandler.kt
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/hybrid/HybridRouterHandler.kt:58]: I think this one needs to stay - if the primary route fails but the fallback succeeds, developers would never know that primary failed. This log is the only way to get this communicated.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/ElapsedTime.kt
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/ElapsedTime.kt:19]: Removed altogether, as we discussed last time it's not used.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: Opted to remove.

**Comment by LukasPaczos** [comment]
Ready for another round.

**Review by korshaknn — APPROVED** [review]
LGTM

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt
Comment (RingerJK) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt:6]: add pls `todo move to mapbox-java`

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt
Comment (RingerJK) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt:9]: could you add doc that String should be URL

**Review by Guardiola31337 — APPROVED** [review]
Thanks for addressing the feedback :shipit:

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt:6]: I'll follow up separately to not block the PR.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt:9]: I'll follow up separately to not block the PR.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt
Comment (danpat) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt]: It should never occur - if it does, it's a server-side bug.

**Pull request merged**
